File: apps/quest-unity-client/Assets/QuestPhoneStream/Scripts/SpatialProtocol.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace QuestPhoneStream
{
    [Serializable]
    public sealed class SpatialDeviceDescriptor
    {
        public string id;
        public string name;
        public string[] protocolVersions;
    }

    [Serializable]
    public sealed class SpatialPayload
    {
        public string[] supportedVersions;
        public string selectedVersion;
        public SpatialDeviceDescriptor device;
        public SpatialCapabilityDescriptor[] capabilities;
        public string code;
        public string message;
        public bool retryable;
        public string subscriptionId;
        public string capability;
        public float rateHz;
        public string format;
        public string transport;
        public string reliability;
    }

    [Serializable]
    public sealed class SpatialEnvelope
    {
        public string v;
        public string id;
        public string type;
        public string source;
        public string target;
        public string sessionId;
        public string streamId;
        public string correlationId;
        public long timestamp;
        public SpatialPayload payload;
    }

    [Serializable]
    public sealed class SpatialVector3
    {
        public float x;
        public float y;
        public float z;
    }

    [Serializable]
    public sealed class SpatialQuaternion
    {
        public float x;
        public float y;
        public float z;
        public float w;
    }

    [Serializable]
    public sealed class SpatialPose
    {
        public string space;
        public long timestamp;
        public SpatialVector3 position;
        public SpatialQuaternion orientation;
    }

    [Serializable]
    public sealed class SpatialTransform
    {
        public string spaceFrom;
        public string spaceTo;
        public long timestamp;
        public SpatialVector3 translation;
        public SpatialQuaternion rotation;
    }

    public static class SpatialWire
    {
        public const string Version = "1.0";
        private static readonly HashSet<string> MessageTypes = new HashSet<string>(StringComparer.Ordinal)
        {
            "device.hello",
            "device.capabilities.get",
            "device.capabilities.result",
            "device.capabilities.changed",
            "subscription.create",
            "subscription.created",
            "subscription.cancel",
            "subscription.closed",
            "protocol.error"
        };

        public static SpatialEnvelope Create(
            string type,
            string source,
            string target,
            SpatialPayload payload = null,
            string sessionId = "",
            string streamId = "",
            string correlationId = "")
        {
            if (!MessageTypes.Contains(type)) throw new ArgumentException("Unsupported Spatial message type");
            return new SpatialEnvelope
            {
                v = Version,
                id = Guid.NewGuid().ToString("N"),
                type = type,
                source = source,
                target = target,
                sessionId = sessionId ?? string.Empty,
                streamId = streamId ?? string.Empty,
                correlationId = correlationId ?? string.Empty,
                timestamp = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                payload = payload ?? new SpatialPayload()
            };
        }

        public static bool TryParse(string json, out SpatialEnvelope envelope)
        {
            envelope = null;
            if (string.IsNullOrWhiteSpace(json)) return false;
            try { envelope = JsonUtility.FromJson<SpatialEnvelope>(json); }
            catch (Exception) { return false; }
            return envelope != null && envelope.v == Version && MessageTypes.Contains(envelope.type) &&
                   !string.IsNullOrWhiteSpace(envelope.id) && !string.IsNullOrWhiteSpace(envelope.source) &&
                   !string.IsNullOrWhiteSpace(envelope.target) && envelope.payload != null;
        }

        public static string Serialize(SpatialEnvelope envelope) => JsonUtility.ToJson(envelope);

        public static string NegotiateVersion(string[] offered)
        {
            if (offered == null) return null;
            foreach (var version in offered) if (version == Version) return Version;
            return null;
        }

        public static SpatialPayload HelloPayload(string deviceId, string selectedVersion = null)
        {
            return new SpatialPayload
            {
                supportedVersions = selectedVersion == null ? new[] { Version } : null,
                selectedVersion = selectedVersion,
                device = new SpatialDeviceDescriptor { id = deviceId, name = deviceId, protocolVersions = new[] { Version } }
            };
        }

        public static SpatialPayload CapabilitiesPayload(SpatialCapabilityDescriptor[] capabilities) =>
            new SpatialPayload { capabilities = capabilities ?? Array.Empty<SpatialCapabilityDescriptor>() };

        public static SpatialPayload ErrorPayload(string code, string message) =>
            new SpatialPayload { code = code, message = message, retryable = false };
    }
}

File: apps/android-agent/app/src/main/java/com/questphonestream/agent/SpatialProtocol.kt
package com.questphonestream.agent

import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

internal object SpatialProtocol {
    const val VERSION = "1.0"
    val MESSAGE_TYPES = setOf(
        "device.hello",
        "device.capabilities.get",
        "device.capabilities.result",
        "device.capabilities.changed",
        "subscription.create",
        "subscription.created",
        "subscription.cancel",
        "subscription.closed",
        "protocol.error"
    )

    fun isSpatialType(type: String): Boolean = type in MESSAGE_TYPES

    fun validateEnvelope(message: JSONObject): String? {
        if (message.optString("v") != VERSION) return "unsupported_spatial_version"
        if (!isSpatialType(message.optString("type"))) return "unknown_spatial_type"
        for (key in listOf("id", "source", "target", "sessionId", "streamId", "correlationId")) {
            if (!message.has(key) || message.opt(key) !is String) return "invalid_$key"
        }
        if (message.optString("id").isEmpty() || message.optString("source").isEmpty() || message.optString("target").isEmpty())
            return "invalid_spatial_identity"
        if (!message.has("timestamp") || message.opt("timestamp") !is Number) return "invalid_timestamp"
        if (message.optJSONObject("payload") == null) return "invalid_payload"
        return null
    }

    fun envelope(
        type: String,
        source: String,
        target: String,
        payload: JSONObject = JSONObject(),
        sessionId: String = "",
        streamId: String = "",
        correlationId: String = ""
    ): JSONObject = JSONObject()
        .put("v", VERSION)
        .put("id", UUID.randomUUID().toString())
        .put("type", type)
        .put("source", source)
        .put("target", target)
        .put("sessionId", sessionId)
        .put("streamId", streamId)
        .put("correlationId", correlationId)
        .put("timestamp", System.currentTimeMillis())
        .put("payload", payload)

    fun helloPayload(deviceId: String, selectedVersion: String? = null): JSONObject = JSONObject().apply {
        if (selectedVersion == null) put("supportedVersions", JSONArray().put(VERSION))
        else put("selectedVersion", selectedVersion)
        put("device", JSONObject()
            .put("id", deviceId)
            .put("name", deviceId)
            .put("protocolVersions", JSONArray().put(VERSION)))
    }

    fun capabilitiesPayload(capabilities: List<CapabilityDescriptor>): JSONObject = JSONObject()
        .put("capabilities", JSONArray().apply { capabilities.forEach { put(capabilityToJson(it)) } })

    fun negotiateVersion(payload: JSONObject): String? {
        val selected = payload.optString("selectedVersion")
        if (selected.isNotEmpty()) return if (selected == VERSION) VERSION else null
        val offered = payload.optJSONArray("supportedVersions") ?: return null
        for (index in 0 until offered.length()) if (offered.optString(index) == VERSION) return VERSION
        return null
    }

    fun errorPayload(code: String, message: String, retryable: Boolean = false): JSONObject = JSONObject()
        .put("code", code)
        .put("message", message)
        .put("retryable", retryable)

    private fun capabilityToJson(capability: CapabilityDescriptor): JSONObject = JSONObject()
        .put("name", capability.name)
        .put("version", capability.version)
        .put("state", JSONObject()
            .put("available", capability.state.available)
            .put("authorized", capability.state.authorized)
            .put("active", capability.state.active))
        .put("transports", JSONArray(capability.transports))
        .put("features", JSONArray(capability.features))
        .put("limits", JSONArray().apply {
            capability.limits.forEach { put(JSONObject().put("name", it.name).put("value", it.value)) }
        })
        .put("permissions", JSONArray(capability.permissions))
}
